### DISCUSSION TITLE
[Proposal] public internal

### DISCUSSION THREAD

**Discussion by JustNrik** [dotnet/csharplang#2283] [Language Ideas] [answered]
Public exposure, Internal consumption. Classes and Interfaces with this keyword combination are exposed as public members but can't be inherited/implemented outside the assembly, they are for internal consumption only.

**Comment by Joe4evr** [comment]
So basically, you want [`InternalImplementationOnly`](https://github.com/dotnet/csharplang/blob/master/meetings/2015/LDM-2015-02-04.md#1-internal-implementation-only).

**Comment by JustNrik** [comment]
Yes

**Comment by DavidArno** [comment]
Difficult to be sure here, as DIMs in C# 8 are still a bit of an unknown, but in theory, 
```cs
public IInternalImplementationOnly
{
    internal void InternalOnlyImplementationGuard();
}
```
should cover this already as only internal classes can implement `InternalOnlyImplementationGuard`. Until we get a release of VS219 that includes DIMs though, it's hard to known if that will really work.

**Comment by jnm2** [comment, marked as answer]
Duplicate of https://github.com/dotnet/csharplang/issues/485, https://github.com/dotnet/csharplang/issues/1716, https://github.com/dotnet/csharplang/issues/1790.

**Comment by JustNrik** [comment]
It looks like java 15 is implementing some feature similar to this, sealed classes/interfaces, sealed here meaning that these classes/interfaces can only be inherited/implemented by the types specified by `permits` keyword. It would be interesting to see how would that work here or be implemented.

I would proprose it as `closed`/`allows`.
```cs
public closed interface IFoo
    allows Foo
{
    void Bar();
}
```

So the interface is publicy exposed, but can only be implemented by Foo, no other class/type can implement it. I'm not sure how would this work with interface inheritance but I would guess they would inherit the same restriction or would it have to be explicit? What do you think guys?
